src/canard.c

```diff
@@ -1,6 +1,736 @@
+/*
+ * Copyright (C) 2015 Michael Sierra <sierramichael.a@gmail.com>
+ */
+
 #include "canard.h"
 #include "canard_internals.h"
 
+#define CANARD_MAKE_TRANSFER_DESCRIPTOR(data_type_id, transfer_type, \
+                                        src_node_id, dst_node_id) \
+    ((data_type_id) | ((transfer_type) << 16) | \
+     ((src_node_id) << 18) | ((dst_node_id) << 25))
+
+struct CanardTxQueueItem
+{
+  CanardTxQueueItem* next;
+  CanardCANFrame frame;
+};
+
+/**
+ *  API functions
+ */
+
+/**
+ * Initializes the library state.
+ * Local node ID will be set to zero, i.e. the node will be anonymous.
+ */
+void canardInit(CanardInstance* out_ins, canardOnTransferReception on_reception)
+{
+  out_ins->node_id = CANARD_BROADCAST_NODE_ID;
+  out_ins->on_reception = on_reception;
+  out_ins->rx_states = NULL;
+  out_ins->tx_queue = NULL;
+  canardInitPoolAllocator(&out_ins->allocator, out_ins->buffer, CANARD_AVAILABLE_BLOCKS);
+}
+
+/**
+ * Assigns a new node ID value to the current node.
+ */
+void canardSetLocalNodeID(CanardInstance* ins, uint8_t self_node_id)
+{
+  ins->node_id = self_node_id;
+}
+
+/**
+ * Returns node ID of the local node.
+ * Returns zero if the node ID has not been set.
+ */
+uint8_t canardGetLocalNodeID(const CanardInstance* ins)
+{
+  return ins->node_id;
+}
+
+/**
+ * Sends a broadcast transfer.
+ * If the node is in passive mode, only single frame transfers will be allowed.
+ */
+ int canardBroadcast(CanardInstance* ins,
+                      uint16_t data_type_id,
+                      uint8_t* inout_transfer_id,
+                      uint8_t priority,
+                      const void* payload,
+                      uint16_t payload_len)
+ {
+  if (payload == NULL)
+  {
+      return -1;
+  }
+  if (canardGetLocalNodeID(ins) == 0 && payload_len > 7)
+  {
+    return -1;
+  }
+  if (priority > 31)
+  {
+      return -1;
+  }
+
+  const uint32_t can_id = ((uint32_t)priority << 24) | ((uint32_t)data_type_id << 8) | (uint32_t)canardGetLocalNodeID(ins);
```

Comment:
Here, if the local node ID is zero, we need to initialize the discriminator field with (pseudo)random bits. A good way to get pseudorandom bits is to apply the transfer CRC function to the payload and then use any 14 bits of the result (libuavcan works this way).

* Explanation: http://uavcan.org/Specification/4._CAN_bus_transport_layer/#discriminator
* Implementation in libuavcan: https://github.com/UAVCAN/libuavcan/blob/master/libuavcan/src/transport/uc_frame.cpp#L171-L176